// socketioauth/index.js
var _            = require('underscore'),
    Express      = require('express'),
    Connect      = require('connect'),
    SocketIo     = require('socket.io'),
    MongoJs      = require('mongojs'),

    OauthReq     = require('./lib/oauthreq'),
    Class        = require('./lib/class'),

    MemoryStore  = Express.session.MemoryStore;

module.exports = Class.extend({
    config:    {},
    store:     null,
    server:    null,
    iosockets: null,
    db:        null,

    init: function(config) {
        console.log('Starting SocketIOAuth');

        // Save Configs for later
        this.config = config;

        if (this.config.db) {
            // Setup database connection
            this.db = MongoJs.connect(this.config.db.database || null, this.config.db.collections || null);
        }

        if (this.config.server) {
            // The server provides access to Express server
            this.server = Express.createServer();
            this.store  = new MemoryStore();

            // * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
            // Create a CONNECT server, add routes for a main page to start chat and a callback
            //
            // On INDEX: if no user in session, get Tumblr authorization routed to /callback
            //           otherwise, start chat server based on user name
            //
            // On CALLBACK: Authenticate with Tumblr, parse XML, store user in a session
            //
            this.server.configure(_.bind(function () {
                this.server.use(Express.cookieParser());
                this.server.use(Express.session({
                    store:  this.store,
                    secret: this.config.server.secret || 'sauce',
                    key:    'express.sid'
                }));

                if (this.config.server.publicPath) {
                    // Configure all public paths to be available (css, img, js, etc)
                    this.server.use(Express.static(this.config.server.publicPath));
                }

                if (this.config.oauth) {
                    // Attach server requests for OAuth automation
                    OauthReq(this.server, this.config.oauth);
                }

                this.server.listen(this.config.server.port || process.env.PORT, this.config.server.host || '0.0.0.0');
            }, this));

            if (this.config.socketio) {
                // Iosockets is the top level connection to Socket.IO
                this.iosockets = SocketIo.listen(this.server);

                // Link memory stores between express and socket.io
                this.iosockets.set('authorization', _.bind(function (data, accept) {
                    if (data.headers.cookie) {
                        data.cookie = Connect.utils.parseCookie(data.headers.cookie);
                        data.sessionID = data.cookie['express.sid'];
                        // (literally) get the session data from the session store
                        this.store.get(data.sessionID, function (err, session) {
                            if (err) {
                                // if we cannot grab a session, turn down the connection
                                accept(err.message, false);
                            } else {
                                // save the session data and accept the connection
                                data.session = session;
                                accept(null, true);
                            }
                        });
                    } else {
                       return accept('No cookie transmitted.', false);
                    }
                }, this));
            }


        }


        // Let user set log level for socket info
        this.iosockets.set('log level', config.socketio.logLevel || 2);

        // Bind the event for client connection (can be overridden by child)
        this.iosockets.on('connection', _.bind(this.onConnect, this));
    },

    onConnect: function(socket)
    {
        socket.on('message', _.bind(this.onMessage, this));
        socket.on('disconnect', _.bind(this.onDisconnect, this));
    },

    onDisconnect: function()
    {
        // Nothing to do right now but tear down
    },

    onMessage: function(message)
    {
        // Nothing to do right now
    },

    getSession: function(socket)
    {
        return socket.handshake.session || null;
    }
});
